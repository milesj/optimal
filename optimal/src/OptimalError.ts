import { ValidationError } from './ValidationError';

export class OptimalError extends Error {
	errors: ValidationError[] = [];

	file: string = '';

	schema: string = '';

	constructor() {
		super('');

		this.name = 'OptimalError';
	}
}
