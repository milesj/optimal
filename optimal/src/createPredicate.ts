import { AnySchema, InferSchemaType, Predicate } from './types';

/**
 * Create a reusable predicate from a schema that validates a value to
 * return a boolean if no validation errors arise. If null or undefined
 * is returned, will return `false`, regardless of nullable or undefinable
 * state.
 */
export function createPredicate<S extends AnySchema, T = InferSchemaType<S>>(
	schema: S,
): Predicate<T> {
	return (value) => {
		try {
			if (value == null) {
				throw new Error('Avoid null/undefined');
			}

			schema.validate(value);
		} catch {
			return false;
		}

		return true;
	};
}
