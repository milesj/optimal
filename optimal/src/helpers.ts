import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
