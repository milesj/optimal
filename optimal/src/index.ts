/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import * as schemas from './schemas';

export type Schemas = typeof schemas;

export { schemas };
export * from './createPredicate';
export * from './createSchema';
export * from './criteria';
export * from './optimal';
export * from './OptimalError';
export * from './schemas/array';
export * from './schemas/blueprint';
export * from './schemas/bool';
export * from './schemas/custom';
export * from './schemas/date';
export * from './schemas/func';
export * from './schemas/instance';
export * from './schemas/lazy';
export * from './schemas/number';
export * from './schemas/object';
export * from './schemas/regex';
export * from './schemas/schema';
export * from './schemas/shape';
export * from './schemas/string';
export * from './schemas/tuple';
export * from './schemas/union';
export * from './types';
export * from './ValidationError';
