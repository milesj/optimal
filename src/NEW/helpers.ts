import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
