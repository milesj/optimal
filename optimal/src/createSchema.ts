import { invalid, tryAndCollect } from './helpers';
import { OptimalError } from './OptimalError';
import {
	AnySchema,
	Criteria,
	CriteriaFactory,
	DefaultValueInitializer,
	InferSchemaType,
	Schema,
	SchemaOptions,
	SchemaState,
	SchemaValidateOptions,
} from './types';

/**
 * Run all validation checks that have been enqueued and return a type casted value.
 * If a value is undefined, inherit the default value, else throw if required.
 * If nullable and the value is null, return early.
 */
function validate<T>(
	state: SchemaState<T>,
	validators: Criteria<T>[],
	initialValue: unknown,
	path: string = '',
	{
		collectErrors = true,
		currentObject = {},
		rootObject = currentObject,
	}: SchemaValidateOptions = {},
): T | null {
	const { defaultValue, metadata } = state;

	let value: unknown = initialValue;

	// Handle undefined
	if (value === undefined) {
		value =
			typeof defaultValue === 'function'
				? (defaultValue as DefaultValueInitializer<T>)(path, currentObject, rootObject)
				: defaultValue;

		invalid(!state.required, 'Field is required and must be defined.', path);
	} else {
		if (__DEV__ && metadata.deprecatedMessage) {
			// eslint-disable-next-line no-console
			console.info(`Field "${path}" is deprecated. ${metadata.deprecatedMessage}`);
		}

		invalid(!state.never, 'Field should never be used.', path);
	}

	// Handle null
	if (value === null) {
		invalid(state.nullable, 'Null is not allowed.', path);
	}

	// Run validations and produce a new value
	const optimalError = new OptimalError();

	validators.forEach((test) => {
		if (
			(test.skipIfNull && value === null) ||
			(test.skipIfOptional && !state.required && value === state.defaultValue)
		) {
			return;
		}

		tryAndCollect(
			() => {
				const result = test.validate(value as T, path, {
					collectErrors,
					currentObject,
					rootObject,
				});

				if (result !== undefined) {
					value = result as T;
				}
			},
			optimalError,
			collectErrors,
		);
	});

	if (optimalError.errors.length > 0) {
		throw optimalError;
	}

	return value as T;
}

export function createSchema<S extends AnySchema, T = InferSchemaType<S>>(
	{ api, cast, defaultValue, type }: SchemaOptions<T>,
	criteria: (Criteria<T> | CriteriaFactory<T>)[] = [],
): S {
	const state: SchemaState<T> = {
		defaultValue,
		metadata: {},
		never: false,
		nullable: false,
		required: false,
		type,
	};

	const validators: Criteria<T>[] = [];

	criteria.forEach((crit) => {
		if (typeof crit === 'function') {
			const validator = crit(state);

			if (validator) {
				validators.push(validator);
			}
		} else {
			validators.push(crit);
		}
	});

	const schema: Schema<T> = {
		schema() {
			return type;
		},
		type() {
			return state.type;
		},
		validate(value, path, options) {
			const result = validate(state, validators, value, path, options)!;

			return cast && result !== null ? cast(result) : result;
		},
	};

	Object.entries(api).forEach(([name, method]) => {
		Object.defineProperty(schema, name, {
			enumerable: true,
			// Create a new schema so that our chainable API is immutable
			value: (...args: unknown[]) =>
				createSchema({ api, cast, defaultValue, type }, [
					...criteria,
					(nextState) => method(nextState, ...args),
				]),
		});
	});

	return schema as S;
}
