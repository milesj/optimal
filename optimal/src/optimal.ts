import { isObject } from './helpers';
import { OptimalError } from './OptimalError';
import { shape } from './schemas/shape';
import { Blueprint, DeepPartial, SchemaValidateOptions, UnknownObject } from './types';
import { ValidationError } from './ValidationError';

export interface OptimalOptions {
	file?: string;
	name?: string;
	prefix?: string;
	unknown?: boolean;
}

export interface Optimal<T extends object> {
	configure: (options: OptimalOptions) => void;
	validate: (struct: DeepPartial<T>, options?: SchemaValidateOptions) => Required<T>;
}

export function optimal<Schemas extends object>(
	blueprint: Blueprint<Schemas>,
	baseOpts: OptimalOptions = {},
): Optimal<Schemas> {
	const options: OptimalOptions = {};
	let schema = shape(blueprint);

	function configure(nextOpts: OptimalOptions) {
		if (!isObject(nextOpts)) {
			throw new TypeError('Optimal options must be a plain object.');
		}

		Object.assign(options, nextOpts);
		schema = schema.exact(!options.unknown);
	}

	configure(baseOpts);

	return {
		configure,
		validate(struct, validateOptions) {
			const object = struct as UnknownObject;

			try {
				return schema.validate(struct, options.prefix ?? '', {
					...validateOptions,
					currentObject: object,
					rootObject: object,
				});
			} catch (error: unknown) {
				let invalid: OptimalError;
				let prefix = '';

				if (error instanceof OptimalError) {
					invalid = error;
				} else {
					invalid = new OptimalError();
					invalid.addError(error as ValidationError);
				}

				if (options.name) {
					invalid.schema = options.name;
					prefix = options.name;
				}

				if (options.file) {
					invalid.file = options.file;
					prefix = prefix ? `${prefix} (${options.file})` : options.file;
				}

				if (prefix) {
					invalid.message = `${prefix}: ${invalid.message}`;
				}

				throw invalid;
			}
		},
	};
}
