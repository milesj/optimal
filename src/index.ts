/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal from './optimal';
import Schema from './Schema';
import Predicate, { custom, func } from './Predicate';
import ArrayPredicate, { array } from './predicates/Array';
import BooleanPredicate, { bool } from './predicates/Boolean';
import InstancePredicate, { instance, date, predicate, regex } from './predicates/Instance';
import NumberPredicate, { number } from './predicates/Number';
import ObjectPredicate, { object, blueprint } from './predicates/Object';
import ShapePredicate, { shape } from './predicates/Shape';
import StringPredicate, { string } from './predicates/String';
import TuplePredicate, { tuple } from './predicates/Tuple';
import UnionPredicate, { union } from './predicates/Union';

export const predicates = {
  array,
  blueprint,
  bool,
  custom,
  date,
  func,
  instance,
  number,
  object,
  predicate,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  array,
  blueprint,
  bool,
  custom,
  date,
  func,
  instance,
  number,
  object,
  predicate,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  Schema,
  Predicate,
  ArrayPredicate,
  BooleanPredicate,
  InstancePredicate,
  NumberPredicate,
  ObjectPredicate,
  ShapePredicate,
  StringPredicate,
  TuplePredicate,
  UnionPredicate,
};

export type Predicates = typeof predicates;

export * from './types';

export default optimal;
