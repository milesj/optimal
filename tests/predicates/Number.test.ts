import { number, NumberPredicate } from '../../src';
import { runChecks, runInProd } from '../helpers';

describe('NumberPredicate', () => {
  let builder: NumberPredicate;

  beforeEach(() => {
    builder = number(123);
  });

  it('sets default value', () => {
    expect(number(456).defaultValue).toBe(456);
  });

  describe('run()', () => {
    it('errors if a non-number value is used', () => {
      expect(() => {
        runChecks(
          number(),
          // @ts-ignore Testing wrong type
          'foo',
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns default value if value is undefined', () => {
      expect(runChecks(number(123))).toEqual(123);
    });

    it('returns default value from factory if value is undefined', () => {
      expect(
        runChecks(
          number(struct => struct.multiplier * 123),
          undefined,
          {
            struct: { multiplier: 3 },
          },
        ),
      ).toEqual(369);
    });

    describe('production', () => {
      it(
        'returns default value if value is undefined',
        runInProd(() => {
          expect(runChecks(builder)).toBe(123);
        }),
      );

      it(
        'returns default value from factory if value is undefined',
        runInProd(() => {
          expect(runChecks(number(() => 456))).toBe(456);
        }),
      );

      it(
        'bypasses checks and returns value',
        runInProd(() => {
          expect(
            runChecks(
              builder,
              // @ts-ignore Test invalid type
              '123',
            ),
          ).toBe(123);
        }),
      );
    });
  });

  describe('between()', () => {
    beforeEach(() => {
      builder.between(0, 5);
    });

    it('errors if min is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.between('foo', 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if max is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.between(0, 'foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value is not a number', () => {
      expect(() => {
        runChecks(
          builder,
          // @ts-ignore Testing wrong type
          'foo',
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if out of range', () => {
      expect(() => {
        runChecks(builder, 10);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if edge of range', () => {
      expect(() => {
        runChecks(builder, 10);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if in range', () => {
      expect(() => {
        runChecks(builder, 3);
      }).not.toThrow();
    });
  });

  describe('between(): inclusive', () => {
    beforeEach(() => {
      builder.between(0, 5, true);
    });

    it('errors if out of range', () => {
      expect(() => {
        runChecks(builder, 10);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if edge of range', () => {
      expect(() => {
        runChecks(builder, 5);
      }).not.toThrow();
    });
  });

  describe('float()', () => {
    beforeEach(() => {
      builder.float();
    });

    it('errors if value is an integer', () => {
      expect(() => {
        runChecks(builder, 123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value is NaN', () => {
      expect(() => {
        runChecks(builder, NaN);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value is a float', () => {
      expect(() => {
        runChecks(builder, 1.23);
      }).not.toThrow();
    });
  });

  describe('gt()', () => {
    beforeEach(() => {
      builder.gt(5);
    });

    it('errors if min is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.gt('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if below minimum', () => {
      expect(() => {
        runChecks(builder, 3);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if edge', () => {
      expect(() => {
        runChecks(builder, 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if above minimum', () => {
      expect(() => {
        runChecks(builder, 10);
      }).not.toThrow();
    });
  });

  describe('gte()', () => {
    beforeEach(() => {
      builder.gte(5);
    });

    it('errors if min is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.gte('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if below minimum', () => {
      expect(() => {
        runChecks(builder, 3);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if edge', () => {
      expect(() => {
        runChecks(builder, 5);
      }).not.toThrow();
    });

    it('doesnt error if above minimum', () => {
      expect(() => {
        runChecks(builder, 10);
      }).not.toThrow();
    });
  });

  describe('int()', () => {
    beforeEach(() => {
      builder.int();
    });

    it('errors if value is a float', () => {
      expect(() => {
        runChecks(builder, 1.23);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value is NaN', () => {
      expect(() => {
        runChecks(builder, NaN);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value is an integer', () => {
      expect(() => {
        runChecks(builder, 123);
      }).not.toThrow();
    });
  });

  describe('lt()', () => {
    beforeEach(() => {
      builder.lt(5);
    });

    it('errors if max is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.lt('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if above maximum', () => {
      expect(() => {
        runChecks(builder, 7);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if edge', () => {
      expect(() => {
        runChecks(builder, 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if below maximum', () => {
      expect(() => {
        runChecks(builder, 3);
      }).not.toThrow();
    });
  });

  describe('lte()', () => {
    beforeEach(() => {
      builder.lte(5);
    });

    it('errors if max is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.lte('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if above maximum', () => {
      expect(() => {
        runChecks(builder, 7);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if edge', () => {
      expect(() => {
        runChecks(builder, 5);
      }).not.toThrow();
    });

    it('doesnt error if below maximum', () => {
      expect(() => {
        runChecks(builder, 3);
      }).not.toThrow();
    });
  });

  describe('negative()', () => {
    beforeEach(() => {
      builder.negative();
    });

    it('errors if value is positive', () => {
      expect(() => {
        runChecks(builder, 1);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value is zero', () => {
      expect(() => {
        runChecks(builder, 1);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value is negative', () => {
      expect(() => {
        runChecks(builder, -1);
      }).not.toThrow();
    });
  });

  describe('oneOf()', () => {
    beforeEach(() => {
      builder.oneOf([123, 456, 789]);
    });

    it('errors if not an array', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.oneOf(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if array is empty', () => {
      expect(() => {
        builder.oneOf([]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if array contains a non-number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.oneOf(['foo', 123]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value is not in the list', () => {
      expect(() => {
        runChecks(builder, 666);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value contains token', () => {
      expect(() => {
        runChecks(builder, 123);
      }).not.toThrow();
    });
  });

  describe('positive()', () => {
    beforeEach(() => {
      builder.positive();
    });

    it('errors if value is negative', () => {
      expect(() => {
        runChecks(builder, -1);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value is zero', () => {
      expect(() => {
        runChecks(builder, 0);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value is positive', () => {
      expect(() => {
        runChecks(builder, 1);
      }).not.toThrow();
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name', () => {
      expect(number().typeAlias()).toBe('number');
    });
  });
});
