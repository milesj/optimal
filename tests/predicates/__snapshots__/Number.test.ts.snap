// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberBuilder between() errors if edge of range 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5."`;

exports[`NumberBuilder between() errors if max is a not a number 1`] = `"Between requires a minimum and maximum number."`;

exports[`NumberBuilder between() errors if min is a not a number 1`] = `"Between requires a minimum and maximum number."`;

exports[`NumberBuilder between() errors if out of range 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5."`;

exports[`NumberBuilder between() errors if value is not a number 1`] = `"Invalid field \\"key\\". Must be a number."`;

exports[`NumberBuilder between(): inclusive errors if out of range 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5 inclusive."`;

exports[`NumberBuilder float() errors if value is NaN 1`] = `"Invalid field \\"key\\". Number must be a float."`;

exports[`NumberBuilder float() errors if value is an integer 1`] = `"Invalid field \\"key\\". Number must be a float."`;

exports[`NumberBuilder gt() errors if below minimum 1`] = `"Invalid field \\"key\\". Number must be greater than 5."`;

exports[`NumberBuilder gt() errors if edge 1`] = `"Invalid field \\"key\\". Number must be greater than 5."`;

exports[`NumberBuilder gt() errors if min is a not a number 1`] = `"Greater-than requires a minimum number."`;

exports[`NumberBuilder gte() errors if below minimum 1`] = `"Invalid field \\"key\\". Number must be greater than or equal to 5."`;

exports[`NumberBuilder gte() errors if min is a not a number 1`] = `"Greater-than requires a minimum number."`;

exports[`NumberBuilder int() errors if value is NaN 1`] = `"Invalid field \\"key\\". Number must be an integer."`;

exports[`NumberBuilder int() errors if value is a float 1`] = `"Invalid field \\"key\\". Number must be an integer."`;

exports[`NumberBuilder lt() errors if above maximum 1`] = `"Invalid field \\"key\\". Number must be less than 5."`;

exports[`NumberBuilder lt() errors if edge 1`] = `"Invalid field \\"key\\". Number must be less than 5."`;

exports[`NumberBuilder lt() errors if max is a not a number 1`] = `"Less-than requires a maximum number."`;

exports[`NumberBuilder lte() errors if above maximum 1`] = `"Invalid field \\"key\\". Number must be less than or equal to 5."`;

exports[`NumberBuilder lte() errors if max is a not a number 1`] = `"Less-than requires a maximum number."`;

exports[`NumberBuilder negative() errors if value is positive 1`] = `"Invalid field \\"key\\". Number must be negative."`;

exports[`NumberBuilder negative() errors if value is zero 1`] = `"Invalid field \\"key\\". Number must be negative."`;

exports[`NumberBuilder oneOf() errors if array contains a non-number 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if array is empty 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if not an array 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if value is not in the list 1`] = `"Invalid field \\"key\\". Number must be one of: 123, 456, 789"`;

exports[`NumberBuilder positive() errors if value is negative 1`] = `"Invalid field \\"key\\". Number must be positive."`;

exports[`NumberBuilder positive() errors if value is zero 1`] = `"Invalid field \\"key\\". Number must be positive."`;

exports[`NumberBuilder run() errors if a non-number value is used 1`] = `"Invalid field \\"key\\". Must be a number."`;
